activerecord/lib/active_record/relation/batches.rb

```diff
@@ -274,46 +292,54 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
           end
         end
 
+        orders_size = orders.size
+        operators = orders.map.with_index(1) do |ord, index|
```

Comment:
I generally prefer to avoid using indexes as they are very primitive and hide richer concepts. For example in our case indexes hide the fact that we want to treat "the last" object in a special way. It would be very helpful if we could write the code with an explicit "last concept"

Assuming that we won't be pre-building the `orders` array in `convert_order_param_to_array` anymore we could do it here like:
```ruby
orders = Array(primary_key).zip(orders)
last_order_value = orders.pop[1]
operators = orders.map do |column, order|
    o == :desc ? :lt : :gt
end
operators << last_order_value  == :desc ? :lteq : :gteq
```

We may need to extract `Array(primary_key).zip(orders)` and do the conversion once as it looks like we are using "orders by pk column" mapping in several places

Reply:
#47901

Reply:
(#47901)-